### PR TITLE
Introduce exception handling information

### PR DESCRIPTION
**Pull request by chuigda**
This is the pull request for #13

### CODE REVIEW

File: pr47-core/src/vm/al31f/executor/async_executor.rs [hunk 1]
```diff
@@ -40,6 +42,40 @@ pub async fn create_vm_main_thread<A: Alloc>(
...
+    let func_id: usize = stack_frame.func_id;
+    let compiled_function: &CompiledFunction = &program.functions[func_id];
+    if let Some(exc_handlers /*: &Box<[ExceptionHandlingBlock]>*/)
+        = &compiled_function.exc_handlers
+    {
+        for exc_handler /*: &ExceptionHandlingBlock*/ in exc_handlers.as_ref().iter() {
+            let (start_insc, end_insc): (usize, usize) = exc_handler.insc_ptr_range;
+            if insc_ptr >= start_insc
+                && insc_ptr <= end_insc
+                && /* exc_handler.exception_id == checked_exception.type_id() */ false {
```
Comment (chuigda) [inline pr47-core/src/vm/al31f/executor/async_executor.rs]: By this time [`std::error::Error::type_id`](https://github.com/rust-lang/rust/issues/60784) is still unstable. We need some other way to achieve this stuff.

Comment (chuigda) [reply pr47-core/src/vm/al31f/executor/async_executor.rs]: **[MID]** Perhaps we can make a `throw!` macro for creating a checked exception. Or maybe a macro is not necessary, just using functions will be fine?

File: pr47-core/src/data/exception.rs [hunk 2]
```diff
@@ -7,7 +7,9 @@ pub enum UncheckedException {
     InvalidBinaryOp { bin_op: char, lhs: Value, rhs: Value, reason: String }
 }
 
+pub type CheckedException = Box<dyn 'static + Error + Send>;
```
Comment (CousinZe) [inline pr47-core/src/data/exception.rs]: If an exception (an `Error` type) is going to be used in by Pr47 program, then it must satisfy `StaticBase` somehow, and can be type-checked and casted at run time. I suggest a Box<dyn DynBase>.

Comment (chuigda) [reply pr47-core/src/data/exception.rs]: **[MID]** Then, here comes the question: for the `Error`s that do not satisfy `StaticBase`, user may hope them to be `UncheckedException`s. That makes sense I think. However we don't have `specialisation` or `min-specialisation` yet.

File: pr47-core/src/vm/al31f/executor/async_executor.rs [hunk 3]
```diff
@@ -265,7 +318,37 @@ pub async unsafe fn vm_thread_run_function<A: Alloc>(
                     return Ok(ret_vec);
                 }
             }
-            Insc::FFICallTyck(_, _, _) => {}
+            Insc::FFICallTyck(ffi_func_id, args, ret_value_locs) => {
```
Comment (CousinZe) [inline pr47-core/src/vm/al31f/executor/async_executor.rs:321]: Are you sure? I think testing with something like `Raise` could be better.

Comment (chuigda) [reply pr47-core/src/vm/al31f/executor/async_executor.rs:321]: Acknowledged

File: pr47-core/src/data/exception.rs [hunk 4]
```diff
@@ -1,13 +1,23 @@
...
     InvalidBinaryOp { bin_op: char, lhs: Value, rhs: Value, reason: String }
 }
 
+#[cfg(feature = "async")]
+pub type UserUncheckedException = Box<dyn Error + Send + 'static>;
+
+#[cfg(not(feature = "async"))]
+pub type UserUncheckedException = Box<dyn Error + 'static>;
+
+pub type CheckedException = Box<dyn DynBase>;
```
Comment (CousinZe) [inline pr47-core/src/data/exception.rs]: Here comes the question. A `CheckedException` may be created by `raise` or `throw` in Pr47 codes. That exceptions will be in the form of  `Value`s, managed by VM heap. Box<dyn DynBase> might not be adequate.

Comment (chuigda) [reply pr47-core/src/data/exception.rs]: Using `Wrapper` to wrap concrete exception up will be just fine, I guess.

Comment (CousinZe) [reply pr47-core/src/data/exception.rs]: Then we need to rewrite our current `Exception` type alot, together with our procedural macro designs.

**Comment by chuigda** [comment]
Both miri test and valgrind test passing, this should be fine. @SchrodingerZhu @PragmaTwice review requested.

**Comment by SchrodingerZhu** [comment]
lgtm

**Pull request merged**